File: src/java/MergeSort.java
// DSA16 (Java): Merge Sort

import java.util.Scanner;

class MergeSort {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the number of elements: ");
        int n = sc.nextInt();

        int[] arr = new int[n];
        System.out.print("Enter the elements: ");
        for (int i = 0; i < n; i++)
            arr[i] = sc.nextInt();

        System.out.print("The array elements before sorting are: ");
        for (int i : arr)
            System.out.print(i + " ");
        System.out.println();

        mergeSort(arr, 0, arr.length - 1);

        System.out.print("The array elements after sorting are: ");
        for (int i : arr)
            System.out.print(i + " ");
        System.out.println();

        sc.close();
    }

    static void mergeSort(int[] arr, int left, int right) {
        if (left >= right)
            return;

        int mid = left + (right - left) / 2;
        
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);

        merge(arr, left, mid, right);
    }

    static void merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] l = new int[n1];
        int[] r = new int[n2];

        for (int i = 0; i < n1; i++)
            l[i] = arr[left + i];

        for (int i = 0; i < n2; i++)
            r[i] = arr[mid + 1 + i];

        int i = 0;
        int j = 0;
        int k = left;

        while (i < n1 && j < n2) {
            if (l[i] <= r[j])
                arr[k++] = l[i++];
            else
                arr[k++] = r[j++];
        }

        while (i < n1)
            arr[k++] = l[i++];

        while (j < n2)
            arr[k++] = r[j++];
    }
}

File: src/kotlin/MergeSort.kt
// DSA16 (Kotlin): Merge Sort

import java.util.Scanner

fun main() {
    val sc = Scanner(System.`in`)

    print("Enter the number of elements: ")
    val n = sc.nextInt()

    val arr = IntArray(n)
    print("Enter the elements: ")
    for (i in arr.indices)
        arr[i] = sc.nextInt()

    print("The array elements before sorting are: ")
    for (i in arr)
        print("$i ")
    println()

    mergeSort(arr, 0, arr.size - 1)

    print("The array elements after sorting are: ")
    for (i in arr)
        print("$i ")
    println()

    sc.close()
}

fun mergeSort(arr: IntArray, left: Int, right: Int) {
    if (left >= right)
        return

    val mid = left + (right - left) / 2
    
    mergeSort(arr, left, mid)
    mergeSort(arr, mid + 1, right)

    merge(arr, left, mid, right)
}

fun merge(arr: IntArray, left: Int, mid: Int, right: Int) {
    val n1 = mid - left + 1
    val n2 = right - mid

    val l = IntArray(n1)
    val r = IntArray(n2)

    for (i in 0 until n1)
        l[i] = arr[left + i]

    for (i in 0 until n2)
        r[i] = arr[mid + 1 + i]

    var i = 0
    var j = 0
    var k = left

    while (i < n1 && j < n2) {
        if (l[i] <= r[j])
            arr[k++] = l[i++]
        else
            arr[k++] = r[j++]
    }

    while (i < n1)
        arr[k++] = l[i++]

    while (j < n2)
        arr[k++] = r[j++]
}
